fix(common): canceled JSONP requests won't throw error with missing callback function

martinsik opened the pull request:
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Fixes #34818

Unsubscribing from a JSONP request will throw console errors such as `Uncaught ReferenceError: ng_jsonp_callback_41 is not defined`. This is happening because the callback function is removed right on unsubscription but this doesn't stop the JSONP script from loading.

https://stackblitz.com/edit/angular-4zmfjr?file=src/app/app.component.ts

## What is the new behavior?

Unsubscribing from JSONP request won't remove callback handler. Handler is then removed by the handler itself:

https://github.com/angular/angular/blob/2365bb89d7a37089f596705db19b1ee5d0366b42/packages/common/http/src/jsonp.ts#L107

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

martinsik committed:
fix(common): canceled JSONP requests won't throw console error with missing callback function

This commit fixes a use-case where unsubscribing from a JSONP request will result in "Uncaught ReferenceError: ng_jsonp_callback_xy is not defined"
thrown into console. Unsubscribing won't remove its associated callback function because the requested script will finish
loading anyway and will try to call the handler.

PR Close #34818

martinsik commented on packages/common/http/test/jsonp_spec.ts:
```diff
@@ -18,7 +18,6 @@ function runOnlyCallback(home: any, data: Object) {
   const keys = Object.keys(home);
   expect(keys.length).toBe(1);
   const callback = home[keys[0]];
-  delete home[keys[0]];
```
I removed this line because the handler should be removed by the handler itself:

https://github.com/angular/angular/blob/2365bb89d7a37089f596705db19b1ee5d0366b42/packages/common/http/src/jsonp.ts#L107

JoostK reviewed (commented):
It appears to be possible to prevent the script from executing by adopting the script element into another document, as described in https://html.spec.whatwg.org/multipage/scripting.html#execute-the-script-block. So instead of removing the script element as we do now, it could instead be adopted into a dummy document. I think that doing so is beneficial in terms of resources, as it avoids the need to parse and execute the script.

petebacondarwin commented:
@martinsik - would you have time to address the comments from @JoostK?

martinsik committed:
refactor(http): change <script>'s ownerDocument in jsonp teardown
handler

Cancel pending json handler by adopting its <script> element into
another document
(https://html.spec.whatwg.org/multipage/scripting.html#execute-the-script-block)
This way the browser will prevent the script from being parsed and executed.

Fixes #34818

martinsik commented:
It's been so long I've completelly forgot I ever made this PR :).

@JoostK I made a small demo to test if changing <script>'s document will prevent it from exeuction and it does work https://stackblitz.com/edit/js-gs6nph?file=index.js  
I refactored the code a little bit. Seems like one state variable (`cancelled`) could be removed completely and I shouldn't even need to use `removeEventListener` because the events won't be triggered after changing the owner document.

I'm not sure why `ci/circleci: legacy-unit-tests-saucelab`  CICD process is failing but it looks like the same problem is in other PRs as well so it doesn't look like I broke it.

JoostK commented on packages/common/http/test/jsonp_spec.ts:
```diff
@@ -63,6 +62,11 @@ const SAMPLE_REQ = new HttpRequest<never>('JSONP', '/test');
+    it('allows the callback to be invoked when the request is cancelled', () => {
+      backend.handle(SAMPLE_REQ).subscribe().unsubscribe();
+      expect(Object.keys(home).length).toBe(0);
+      expect(document.mock!.ownerDocument).not.toEqual(document);
```
Please add a comment here to indicate why this assertion is relevant. Something like:

```ts
// The script element should have been transferred to a different document to prevent it from executing.
```

JoostK commented on packages/common/http/test/jsonp_spec.ts:
```diff
@@ -63,6 +62,11 @@ const SAMPLE_REQ = new HttpRequest<never>('JSONP', '/test');
     });
+    it('allows the callback to be invoked when the request is cancelled', () => {
+      backend.handle(SAMPLE_REQ).subscribe().unsubscribe();
+      expect(Object.keys(home).length).toBe(0);
```
Is this sufficient to verify that the callback handler wasn't executed? Isn't this check only indicating that cleanup happened?

martinsik replied:
I updated the test to also check that the callback wasn't invoked.

JoostK commented on packages/common/http/test/jsonp_spec.ts:
```diff
@@ -63,6 +62,11 @@ const SAMPLE_REQ = new HttpRequest<never>('JSONP', '/test');
       });
       document.mockError(error);
     });
+    it('allows the callback to be invoked when the request is cancelled', () => {
```
I remember proposing this spec description and looking at it now it no longer makes sense to me (sorry!), as there is no sign of a callback anywhere in this test setup, and I think the callback should not at all be invoked, right?

So here's attempt two for a description:

```ts
    it('prevents the script from executing when the request is cancelled', () => {
```

would that be better?

JoostK reviewed (commented):
Thanks for your continued effort on this PR. It looks good to me; I'll request a review from @alxhub as code owner of fw-http so we can get this landed.

JoostK commented:
Btw don't worry about the Saucelabs failures; they have been _very_ flaky over the last couple of weeks and #44281 should help make it more stable.

martinsik commented:
@JoostK I updated this PR and implemented your comments.

martinsik commented on packages/common/http/src/jsonp.ts:
```diff
@@ -56,6 +56,8 @@ export class JsonpClientBackend implements HttpBackend {
   private readonly resolvedPromise = Promise.resolve();
 
+  private readonly dummyDocument = this.document.implementation.createHTMLDocument();
+
```
@AndrewKushnir I moved it into a separate method. I'm not sure if `removeListeners` isn't confusing because it's not really removing any listeners.

martinsik commented:
@AndrewKushnir I rebased the code and implemented your suggestions.

AndrewKushnir reviewed (approved):
@martinsik the change looks good 👍 

I'd like to ask @JoostK to take another look. After that we can run the necessary tests in Google's codebase and proceed with next steps if everything looks good.

Thank you.

JoostK commented on packages/common/http/src/jsonp.ts:
```diff
@@ -56,6 +56,12 @@ export class JsonpClientBackend implements HttpBackend {
+   * When a pending <script> is unsubscribed we'll move it to this document, so it won't be
+   * executed.
+   */
+  private foreignDocument!: Document;
```
nit: I'd personally move this to become a top-level variable, as this can be safely shared between multiple instances (although unlikely to be relevant except for tests) and it minimizes better.

I'd also request not to use `!` here, as we can easily include `|undefined` in the type to make reads of this field null-safe. Or use `Document|null = null;` which produces more code but has the benefit that reads from this class stay monomorphic. But that wouldn't be a concern if it's a top-level var (nor is this performance sensitive code anyway, I'm mentioning it as a general takeaway).

tldr; I'd declare `let foreignDocument: Document|undefined;` as top-level declaration.

martinsik replied:
@JoostK I moved `foreignDocument` outside of the class, but it seems like I'll need to be using `!` in `foreignDocument!.adoptNode(script);` because `this.document` is of type `any` so TypeScript can't know what type is returned from `createHTMLDocument()` .
I could use typecasting on this `(this.document.implementation as DOMImplementation).createHTMLDocument()` but that's probably not correct because `document` could use a different implementation.

JoostK replied:
Ah, interesting. If we're relying on `createHTMLDocument` to be available, then we might as well do the cast. If we don't want to rely on that then we'd need to account for the situation where `implementaiton` or `createHTMLDocument` is not available as we'd expect, but I don't think there's a reasonable alternative we could use in that case so that seems not viable to me.

martinsik replied:
Ok, I added typecast to `DOMImplementation`.

JoostK commented:
@martinsik thanks, LGTM. Could you please rebase on latest master to get CI to run?

jessicajaniuk commented:
This PR was merged into the repository by commit 909b21aa29c1fe8094ce3f7fb70ddf24668ba155.
